### PR TITLE
Implemented #190

### PR DESCRIPTION
**Pull request by manast**
This PR changes the completed and failed sets to be zsets. This allows us to efficiently get a subset of the completed or failed jobs, which is very useful for pagination on UIs, etc.
Along with this PR there are also some refactoring, for instance I started to move the LUA scripts to separate files, which is much more cleaner and convenient than what we have today.

### CODE REVIEW

**Comment by manast** [comment]
@bradvogel ping :)

**Comment by bradvogel** [comment]
Awesome! Will review shortly

File: lib/scripts/updateDelaySet.lua [hunk 1]
```diff
@@ -0,0 +1,35 @@
+--[[
```
Comment (bradvogel) [inline lib/scripts/updateDelaySet.lua]: thanks for moving these into separate files - much cleaner!

File: lib/queue.js [hunk 2]
```diff
@@ -961,17 +958,6 @@ Queue.prototype.getJobs = function(queueType, type, start, end){
     case 'LIST':
       jobs = this.client.lrange(key, start, end);
       break;
-    case 'SET':
```
Comment (bradvogel) [inline lib/queue.js]: much cleaner!

File: lib/scripts/index.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+var fs = require('fs');
+var path = require('path');
+
+//
+// for some very strange reason, defining scripts with this code results in this error
+// when executing the scripts: ERR value is not an integer or out of range
+//
+module.exports = function(client){
+  loadScripts(client, __dirname);
```
Comment (bradvogel) [inline lib/scripts/index.js]: why is this needed in addition to `loadScriptIfNeeded`?

File: lib/scripts.js [hunk 4]
```diff
@@ -230,11 +238,57 @@ var scripts = {
         }
       });
   },
-  moveToCompleted: function(job, removeOnComplete){
+  _moveToCompleted: function(job, removeOnComplete){
```
Comment (bradvogel) [inline lib/scripts.js]: I don't think this is called anymore

File: lib/scripts.js [hunk 5]
```diff
@@ -28,6 +27,14 @@ function isCommandDefined(client, hash){
   return !!client[hash];
 }
 
+var path = require('path');
+var fs = require('fs');
+function loadScriptIfNeeded(queue, name){
```
Comment (bradvogel) [inline lib/scripts.js]: It'd be nice if `scripts/index.js` exposed this API

File: lib/scripts.js (hunk 4 above)
Comment (bradvogel) [inline lib/scripts.js]: I think `Job.prototype.move()` should be removed to, which would allow us to remove the `move` script (which appears to be redundant).

File: lib/scripts/index.js (hunk 3 above)
Comment (manast) [inline lib/scripts/index.js]: This is what it really should be. I want to load all the scripts at start define them as ioredis commands. However this code does not work, the scripts give ERR when executing them. I have spent several hours trying to figure it out without success. On the other hand, ```loadScriptIfNeeded``` worked well. Ultimately I would like to revisit the code in *index.js* to make it work.

File: lib/scripts.js (hunk 4 above)
Comment (manast) [inline lib/scripts.js]: Yes. I have understood that there is no point in trying to have a generic move function, so I will have specific methods instead.

File: lib/scripts.js (hunk 5 above)
Comment (manast) [inline lib/scripts.js]: yes, you are right. I will move it.

**Comment by manast** [comment]
I closed this PR in favour of this one  https://github.com/OptimalBits/bull/pull/496 which would be merged into 3.0.0 branch instead of master.
